Minor improvments: local images and ability to edit and duplicate cards

bmarne opened the pull request.

bmarne committed:
Now you can add a local image to a card
(e.g. `./image/your_image.png`)

bmarne committed:
remove old regex (without local image)

bmarne committed:
New feature: edit cards' text (not the image)

bmarne committed:
New feature: duplicate cards

bmarne committed:
It should update the card data-text value while exporting results

indigane commented:
Thanks for bringing these up.

Are these changes something you would want merged into the main repo, or are they something you need just for yourself in your fork?

Regardless, I'm interested in your use cases for them, if you're willing to share :)

If you want these merged into the main repo, then at least the card editing and creation/duplication needs to be put behind a setting, similar to [`allowCategoryEditing`](https://github.com/indigane/cardsort/blob/261e15f/common-scripts.js#L8). I would add a `allowCardEditing` flag that allows both editing and duplicating.

bmarne commented:
First of all thank, you for this (free software) sorting card tool.

My idea was to share with you some modifications I made. You may want to merge them in your repo.

The uses cases are very common:

- **Editing cards**: in the initialization phase of the card sorting, while letting users browsing the cards' label and discussing with them their meaning, it is common that users submit some labels that are more meaningful for them. With physical cards, we usually strike the old label and add a new one on the card. It is useful to understand how much our labels are meaningful.
- **Duplicate cards**: in the sorting phase, it happens that users want to drag a card in several groups (more like with a tag, than with a section). After a discussion, it is common to allow the users to take a blank card, to duplicate the label on it, and then sort the same label in several groups.

I will explore how to add settings.

indigane commented:
Ah, thanks for the clarifications. Since duplicating cards has a different purpose than editing cards, my earlier suggestion to place them behind a single settings flag might not be a good idea. It could be that there is no card editing phase, but the host still wants to allow duplicating cards.

indigane commented:
If you do edit the settings, there's one piece of complexity that maybe in hindsight wasn't the best idea. The settings flags are packed into the bits of a single number.

https://github.com/indigane/cardsort/blob/main/common-scripts.js#L3-L6

```js
const settingsFlagsEnum = {
  allowCategoryEditing: 1,
  isRandomized: 2,
};
```

So when adding flags, these need to be powers of 2, for example:

```js
const settingsFlagsEnum = {
  allowCategoryEditing: 1,
  isRandomized: 2,
  allowCardEditing: 4,
  allowCardDuplication: 8,
};
```

bmarne committed:
Add settings fo Card Editing and Duplication

bmarne commented:
I added two settings, I hope it is OK.

bmarne committed:
Improvement of the card edition. Now, old label is shown striked

bmarne committed:
small bug

bmarne committed:
Fixes for old label display (works with share and better export)

bmarne committed:
final fixes for showing old text

bmarne committed:
beautify

bmarne commented:
Some final improvements in order to show the old label when it has been edited (it also work with Export and Share). I think my work on this feature is over.

bmarne committed:
better aspect for edited cards in analysis tool

bmarne committed:
Add an Export button to export a CSV of the analysis

bmarne committed:
Merge edit cards feature with export analysis feature

bmarne committed:
Some bugs solved: Forgotten bits

bmarne committed:
Merge last corrections

bmarne committed:
swap x/y for exports

bmarne committed:
Merge more fixes

bmarne committed:
Missing first column header

bmarne committed:
Import more fixes

bmarne committed:
A null cardCount is a 0

bmarne committed:
fixes

bmarne committed:
temprorary changes to the readme, awaiting for PR acceptance

bmarne committed:
Avoid Failed to execute 'btoa' on 'Window': The string to be encoded contains characters outside of the Latin1 range.

bmarne committed:
Merge branch 'UTF8_btoa'

bmarne committed:
Merge branch 'ExportAnalysisTable'

bmarne commented:
Hi,

Do you agree to merge this PR?

I have [some other changes to propose as a PR](https://github.com/bmarne/cardsort/tree/ExportAnalysisTable). It would be easier to submit them once this first PR has been accepted or rejected.

bmarne committed:
Update README.md

bmarne committed:
Images' size in the readme

bmarne committed:
link to the analysis tool

bmarne committed:
New feature: share analysis URL (you can interrupt your analysis and finish it later)

bmarne committed:
typo in readme

bmarne committed:
some emojis as UI helpers

bmarne committed:
UI: pointer on buttons & more emojis

bmarne committed:
fix: share also renamed categories

bmarne committed:
Add example to the Readme

bmarne committed:
Install and howto in the readme

bmarne committed:
typos

bmarne committed:
typos

bmarne committed:
implements a SVG dendrogram from the CSV data

bmarne committed:
add threshold

bmarne committed:
SVG export!

bmarne committed:
Choice of distance calculation + nomilization

bmarne committed:
alpha warning & visual refinement

bmarne committed:
cleaning

bmarne committed:
Don't reduce to binary data (too much lost)

bmarne committed:
avoid eval()

bmarne committed:
Parameter to choose number of cat. for labels

bmarne committed:
typo

bmarne committed:
typo

bmarne committed:
Using the CSV was not a good idea, we lack info on particpants choices. → back to use the JSON instead of the CSV

bmarne committed:
Some UI refactor on analysis page:
- show the spread processed
- move and rename shared button
- allow to export back the whole json of all processed spreads

bmarne committed:
fix: generated links to spreads were missing the '#' category (unsorted cards)
